rllib/execution/train_ops.py

```diff
@@ -65,6 +65,104 @@ def train_one_step(trainer, train_batch) -> Dict:
+    local_worker = workers.local_worker()
+    policies = local_worker.policies_to_train
+    num_sgd_iter = config.get("sgd_num_iter", 1)
+    sgd_minibatch_size = config.get("sgd_minibatch_size", 0)
+
+    # Collect actual GPU devices to use.
+    if not config.get("num_gpus", 0):
+        _fake_gpus = True
+        num_gpus = 1
+    type_ = "cpu" if _fake_gpus else "gpu"
+    devices = [
+        "/{}:{}".format(type_, 0 if _fake_gpus else i)
+        for i in range(int(math.ceil(num_gpus)))
+    ]
+
+    # Make sure total batch size is dividable by the number of devices.
+    # Batch size per tower.
+    per_device_batch_size = sgd_minibatch_size // len(devices)
+    # Total batch size.
+    batch_size = per_device_batch_size * len(devices)
+    assert batch_size % len(devices) == 0
+    assert batch_size >= len(devices), "Batch size too small!"
+
+    # Handle everything as if multi-agent.
+    train_batch = train_batch.as_multi_agent()
+
+    # Load data into GPUs.
+    load_timer = trainer._timers[LOAD_BATCH_TIMER]
+    with load_timer:
+        num_loaded_samples = {}
+        for policy_id, batch in train_batch.policy_batches.items():
+            # Not a policy-to-train.
+            if policy_id not in policies:
+                continue
+
+            # Decompress SampleBatch, in case some columns are compressed.
+            batch.decompress_if_needed()
+
+            # Load the entire train batch into the Policy's only buffer
+            # (idx=0). Policies only have >1 buffers, if we are training
+            # asynchronously.
+            num_loaded_samples[policy_id] = local_worker.policy_map[
+                policy_id].load_batch_into_buffer(
+                    batch, buffer_index=0)
+
+    # Execute minibatch SGD on loaded data.
+    learn_timer = trainer._timers[LEARN_ON_BATCH_TIMER]
+    with learn_timer:
+        # Use LearnerInfoBuilder as a unified way to build the final
+        # results dict from `learn_on_loaded_batch` call(s).
+        # This makes sure results dicts always have the same structure
+        # no matter the setup (multi-GPU, multi-agent, minibatch SGD,
+        # tf vs torch).
+        learner_info_builder = LearnerInfoBuilder(num_devices=len(devices))
+
+        for policy_id, samples_per_device in num_loaded_samples.items():
+            policy = local_worker.policy_map[policy_id]
+            num_batches = max(
+                1,
+                int(samples_per_device) // int(per_device_batch_size))
+            logger.debug("== sgd epochs for {} ==".format(policy_id))
+            for _ in range(num_sgd_iter):
+                permutation = np.random.permutation(num_batches)
+                for batch_index in range(num_batches):
+                    # Learn on the pre-loaded data in the buffer.
+                    # Note: For minibatch SGD, the data is an offset into
+                    # the pre-loaded entire train batch.
+                    results = policy.learn_on_loaded_batch(
```

Comment:
a stupid question, is learn_on_loaded_batch async? or are we learning on multiple GPUs sequentially?

Reply:
Yes, we learn on all GPUs in parallel. Otherwise, it wouldn't make any sense to have multiple GPUs for the policy :)